File: native/window-helper-addon/src/window_helper.cc
#include <napi.h>
#include "window_manager.h"

// Request accessibility permission
Napi::Value RequestPermission(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  bool result = RequestAccessibilityPermission();
  
  return Napi::Boolean::New(env, result);
}

// Minimize all windows
Napi::Value MinimizeAllWindows(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  try {
    WindowManagerResult result = MinimizeAllWindowsNative();
    
    Napi::Object resultObj = Napi::Object::New(env);
    resultObj.Set("success", Napi::Boolean::New(env, result.success));
    resultObj.Set("minimized", Napi::Number::New(env, result.minimized));
    resultObj.Set("total", Napi::Number::New(env, result.total));
    resultObj.Set("error", Napi::String::New(env, result.error));
    resultObj.Set("errorCode", Napi::String::New(env, result.errorCode));
    
    return resultObj;
  } catch (const std::exception& e) {
    Napi::Error::New(env, e.what()).ThrowAsJavaScriptException();
    return env.Null();
  }
}

// Restore all windows
Napi::Value RestoreAllWindows(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  try {
    WindowManagerResult result = RestoreAllWindowsNative();
    
    Napi::Object resultObj = Napi::Object::New(env);
    resultObj.Set("success", Napi::Boolean::New(env, result.success));
    resultObj.Set("restored", Napi::Number::New(env, result.restored));
    resultObj.Set("total", Napi::Number::New(env, result.total));
    resultObj.Set("error", Napi::String::New(env, result.error));
    resultObj.Set("errorCode", Napi::String::New(env, result.errorCode));
    
    return resultObj;
  } catch (const std::exception& e) {
    Napi::Error::New(env, e.what()).ThrowAsJavaScriptException();
    return env.Null();
  }
}

// Hide all apps
Napi::Value HideAllApps(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  try {
    WindowManagerResult result = HideAllAppsNative();
    
    Napi::Object resultObj = Napi::Object::New(env);
    resultObj.Set("success", Napi::Boolean::New(env, result.success));
    resultObj.Set("error", Napi::String::New(env, result.error));
    resultObj.Set("errorCode", Napi::String::New(env, result.errorCode));
    
    return resultObj;
  } catch (const std::exception& e) {
    Napi::Error::New(env, e.what()).ThrowAsJavaScriptException();
    return env.Null();
  }
}

// Minimize specific app
Napi::Value MinimizeApp(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  if (info.Length() < 1 || !info[0].IsString()) {
    Napi::TypeError::New(env, "Expected app name as string").ThrowAsJavaScriptException();
    return env.Null();
  }
  
  std::string appName = info[0].As<Napi::String>().Utf8Value();
  
  try {
    WindowManagerResult result = MinimizeAppNative(appName);
    
    Napi::Object resultObj = Napi::Object::New(env);
    resultObj.Set("success", Napi::Boolean::New(env, result.success));
    resultObj.Set("minimized", Napi::Number::New(env, result.minimized));
    resultObj.Set("total", Napi::Number::New(env, result.total));
    resultObj.Set("error", Napi::String::New(env, result.error));
    resultObj.Set("errorCode", Napi::String::New(env, result.errorCode));
    
    return resultObj;
  } catch (const std::exception& e) {
    Napi::Error::New(env, e.what()).ThrowAsJavaScriptException();
    return env.Null();
  }
}

// Minimize all windows except specified app
Napi::Value MinimizeAllWindowsExcluding(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  if (info.Length() < 1 || !info[0].IsString()) {
    Napi::TypeError::New(env, "Expected app name as string").ThrowAsJavaScriptException();
    return env.Null();
  }
  
  std::string excludeAppName = info[0].As<Napi::String>().Utf8Value();
  
  try {
    WindowManagerResult result = MinimizeAllWindowsExcludingNative(excludeAppName);
    
    Napi::Object resultObj = Napi::Object::New(env);
    resultObj.Set("success", Napi::Boolean::New(env, result.success));
    resultObj.Set("minimized", Napi::Number::New(env, result.minimized));
    resultObj.Set("total", Napi::Number::New(env, result.total));
    resultObj.Set("error", Napi::String::New(env, result.error));
    resultObj.Set("errorCode", Napi::String::New(env, result.errorCode));
    
    return resultObj;
  } catch (const std::exception& e) {
    Napi::Error::New(env, e.what()).ThrowAsJavaScriptException();
    return env.Null();
  }
}

// Check accessibility permissions
Napi::Value CheckAccessibilityPermissions(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  bool hasPermissions = CheckAccessibilityPermissionsNative();
  
  return Napi::Boolean::New(env, hasPermissions);
}

// Initialize the addon
Napi::Object Init(Napi::Env env, Napi::Object exports) {
  exports.Set(Napi::String::New(env, "requestPermission"),
              Napi::Function::New(env, RequestPermission));
  exports.Set(Napi::String::New(env, "minimizeAllWindows"),
              Napi::Function::New(env, MinimizeAllWindows));
  exports.Set(Napi::String::New(env, "restoreAllWindows"),
              Napi::Function::New(env, RestoreAllWindows));
  exports.Set(Napi::String::New(env, "hideAllApps"),
              Napi::Function::New(env, HideAllApps));
  exports.Set(Napi::String::New(env, "minimizeApp"),
              Napi::Function::New(env, MinimizeApp));
  exports.Set(Napi::String::New(env, "minimizeAllWindowsExcluding"),
              Napi::Function::New(env, MinimizeAllWindowsExcluding));
  exports.Set(Napi::String::New(env, "checkAccessibilityPermissions"),
              Napi::Function::New(env, CheckAccessibilityPermissions));
  
  return exports;
}

NODE_API_MODULE(window_helper, Init)

File: src/main/platform/window-helper.ts
import { join } from 'path';
import { app } from 'electron';
import { logger } from '../logger';

let addonInstance: any = null;

// Lazy load the addon
function getWindowHelper() {
  if (addonInstance) {
    return addonInstance;
  }

  try {
    // Try to load the addon
    // In development, it's in native/window-helper-addon
    // In production, it should be bundled with the app
    const isDev = process.env.NODE_ENV === 'development' || !app.isPackaged;
    
    if (isDev) {
      // In development, require from the addon directory
      const addonPath = join(process.cwd(), 'native/window-helper-addon');
      addonInstance = require(addonPath);
      logger.info('Loaded window helper addon from development path', { path: addonPath });
    } else {
      // In production, the addon should be in the app bundle
      // Try multiple possible locations
      const possiblePaths = [
        join(process.resourcesPath || '', 'native/window-helper-addon'),
        join(app.getAppPath(), 'native/window-helper-addon'),
        join(__dirname, '../../native/window-helper-addon'),
        join(__dirname, '../../../native/window-helper-addon'),
      ];

      let loaded = false;
      for (const path of possiblePaths) {
        try {
          addonInstance = require(path);
          logger.info('Loaded window helper addon from production path', { path });
          loaded = true;
          break;
        } catch (err) {
          // Try next path
          logger.debug('Failed to load addon from path', { path, error: err instanceof Error ? err.message : String(err) });
        }
      }

      if (!loaded) {
        throw new Error('Window helper addon not found in any expected location');
      }
    }

    logger.info('Window helper addon loaded successfully');
    return addonInstance;
  } catch (error) {
    logger.error('Failed to load window helper addon', error instanceof Error ? error : new Error(String(error)), {
      isDev: process.env.NODE_ENV === 'development' || !app.isPackaged,
      resourcesPath: process.resourcesPath,
      appPath: app?.getAppPath(),
      cwd: process.cwd(),
      __dirname,
    });
    throw error;
  }
}

export interface WindowHelperResult {
  success: boolean;
  minimized?: number;
  restored?: number;
  total?: number;
  error: string;
  errorCode: string;
}

export const windowHelper = {
  requestPermission(): boolean {
    try {
      return getWindowHelper().requestPermission();
    } catch (error) {
      logger.error('Failed to request accessibility permission', error instanceof Error ? error : new Error(String(error)));
      return false;
    }
  },

  minimizeAllWindows(): WindowHelperResult {
    try {
      return getWindowHelper().minimizeAllWindows();
    } catch (error: any) {
      logger.error('Failed to minimize all windows', error instanceof Error ? error : new Error(error.message));
      return {
        success: false,
        minimized: 0,
        total: 0,
        error: error.message || 'Unknown error',
        errorCode: 'UNKNOWN_ERROR',
      };
    }
  },

  restoreAllWindows(): WindowHelperResult {
    try {
      return getWindowHelper().restoreAllWindows();
    } catch (error: any) {
      logger.error('Failed to restore all windows', error instanceof Error ? error : new Error(error.message));
      return {
        success: false,
        restored: 0,
        total: 0,
        error: error.message || 'Unknown error',
        errorCode: 'UNKNOWN_ERROR',
      };
    }
  },

  hideAllApps(): WindowHelperResult {
    try {
      return getWindowHelper().hideAllApps();
    } catch (error: any) {
      logger.error('Failed to hide all apps', error instanceof Error ? error : new Error(error.message));
      return {
        success: false,
        error: error.message || 'Unknown error',
        errorCode: 'UNKNOWN_ERROR',
      };
    }
  },

  minimizeApp(appName: string): WindowHelperResult {
    try {
      return getWindowHelper().minimizeApp(appName);
    } catch (error: any) {
      logger.error('Failed to minimize app', error instanceof Error ? error : new Error(error.message), { appName });
      return {
        success: false,
        minimized: 0,
        total: 0,
        error: error.message || 'Unknown error',
        errorCode: 'UNKNOWN_ERROR',
      };
    }
  },

  minimizeAllWindowsExcluding(excludeAppName: string): WindowHelperResult {
    try {
      return getWindowHelper().minimizeAllWindowsExcluding(excludeAppName);
    } catch (error: any) {
      logger.error('Failed to minimize all windows excluding app', error instanceof Error ? error : new Error(error.message), { excludeAppName });
      return {
        success: false,
        minimized: 0,
        total: 0,
        error: error.message || 'Unknown error',
        errorCode: 'UNKNOWN_ERROR',
      };
    }
  },

  checkAccessibilityPermissions(): boolean {
    try {
      return getWindowHelper().checkAccessibilityPermissions();
    } catch (error) {
      logger.error('Failed to check accessibility permissions', error instanceof Error ? error : new Error(String(error)));
      return false;
    }
  },
};

